File: typescript/kyu7/count_the_divisors_of_a_number.ts
// 4-- > 3(1, 2, 4);
// 5-- > 2(1, 5);
// 12-- > 6(1, 2, 3, 4, 6, 12);
// 30-- > 8(1, 2, 3, 5, 6, 10, 15, 30);
export function divisors(n: number) {
  let count = 0;
  for (let i = 1; i <= n; i++) {
    if (n % i === 0) {
      count++;
    }
  }
  return count;
}

File: rust/src/kyu7/count_the_divisors_of_a_number.rs
pub fn divisors(n: u32) -> u32 {
    let mut count = 1;
    let mut prime = 2;
    let mut n = n;

    while prime * prime <= n {
        if n % prime != 0 {
            prime += 1;
        } else {
            let mut prime_count = 0;
            while n % prime == 0 {
                n /= prime;
                prime_count += 1;
            }
            count *= prime_count + 1;
        }
    }

    if n > 1 {
        count *= 2;
    }
    count
}

pub fn divisors_2(n: u32) -> u32 {
    (1..=n).filter(|d| n % d == 0).count() as u32
}
